src/run.rs

```diff
@@ -326,23 +328,52 @@ pub fn clean_up(
         }
     } else {
         let _ = fs::remove_file(delete_file);
+        let tmp_dir = preserve_dir(tmp_dir, project_name, preserved_project_dir)?;
 
-        let mut tmp_dir = tmp_dir.into_path();
+        log::info!("Project directory preserved at: {}", tmp_dir.display());
+    }
 
-        if let Some(name) = project_name {
-            let new_path = tmp_dir.with_file_name(name);
+    kill_subprocesses(subprocesses)?;
 
-            fs::rename(&tmp_dir, &new_path)?;
+    Ok(())
+}
 
-            tmp_dir = new_path;
+// preserve dir by renaming to appropriate name if project_name is set
+// and moves the project dir to preserved_project_dir as defined by the user
+// it returns the dir where the project is preserved
+pub fn preserve_dir(
+    tmp_dir: TempDir,
+    project_name: Option<&str>,
+    preserved_project_dir: Option<&str>,
+) -> Result<PathBuf> {
+    let curr_tmp_dir = tmp_dir.into_path();
+    let mut final_tmp_dir = curr_tmp_dir.clone();
+
+    if let Some(name) = project_name {
+        final_tmp_dir = final_tmp_dir.with_file_name(name);
+    }
+
+    if let Some(preserved_project_dir) = preserved_project_dir {
+        let folder_to_move = final_tmp_dir
+            .file_name()
+            .context("cannot find the project's directory name")?;
+        let mut preserved_project_dir = PathBuf::from(preserved_project_dir);
+
+        if !preserved_project_dir.exists() {
+            fs::create_dir_all(&preserved_project_dir)
+                .context("cannot create preserve project's directory")?;
         }
 
-        log::info!("Project directory preserved at: {}", tmp_dir.display());
+        preserved_project_dir.push(folder_to_move);
+
+        final_tmp_dir = preserved_project_dir;
     }
 
-    kill_subprocesses(subprocesses)?;
+    if final_tmp_dir != curr_tmp_dir {
+        fs::rename(&curr_tmp_dir, &final_tmp_dir)?;
+    };
```

Comment:
You should move this in the `let preserve_dir =` block

Reply:
I'm still processing the requests, but from what I see, if I move this to the `let preserve_dir =` block, we wouldn't rename the folder if the user does not want to move the folder (e.g.  `preserved_project_dir` is empty), but the the project was created using `--name my-project`. Which we should rename the folder to it, removing the `tmp-xxxx` part.

Have I missed something?

Reply:
Oh yes my bad! :grimacing: